### PR TITLE
feat: LP v2 migrations

### PR DESCRIPTION
**Pull request by wischli**
# Description

* Removes deprecated `RelayerList` storage
* Migrates deprecated `gateway::DomainRouters` storage to `axelar_router::Configuration` and `axelar_router::ChainNameById`
* Centrifuge (v0 -> v3): Migrates deprecated outbound message queues (normal, failed) to corresponding storages in gateway queue
* Dev (v2 -> v3): Translates queue storage values from `v2::GatewayMessage` to `v3::GatewayMessage`

### Dev
Migration failed here first such that I removed the old entry and re-added it manually.
```
024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Pre checks done!
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::nuke] ClearRelayerList: Cleared all RelayerList storage entries
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::nuke] ClearRelayerList Clear result of RelayerList: backend: 0 unique: 0 loops: 0
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Migration done!
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Post checks done!
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::v2_update_message_queue] InitMessageQueue: Pre checks done with 0 items in v2::MessageQueue, 0 in v3::MessageQueue and 0 items in FailedMessageQueue and 0 in v3::FailedMessageQueue!
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::v2_update_message_queue] InitMessageQueue: Migration done with 0 in total!
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::v2_update_message_queue] InitMessageQueue: Post checks done!
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig PRE Migration counts 1 updated domains and 0 removed domains
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig PRE: Checks done with 1 domains!
[2024-09-03T06:51:38Z INFO  frame_support::migrations] 🚚 Pallet "LiquidityPoolsGateway" VersionedMigration migrating storage version from 2 to 3.
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig ON_RUNTIME_UPGRADE: Migration done with 1 updated domains!
[2024-09-03T06:51:38Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig POST: Checks done!
[2024-09-03T06:51:38Z INFO  runtime::frame-support] ⚠️ Ethereum declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `NoStorageVersionSet`
[2024-09-03T06:51:38Z INFO  runtime::executive] ✅ Entire runtime state decodes without error. 343006 bytes total.
```

### Demo
```
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Pre checks done!
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::nuke] ClearRelayerList: Cleared all RelayerList storage entries
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::nuke] ClearRelayerList Clear result of RelayerList: backend: 13 unique: 13 loops: 13
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Migration done!
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Post checks done!
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::v2_update_message_queue] InitMessageQueue: Pre checks done with 0 items in v2::MessageQueue, 0 in v3::MessageQueue and 0 items in FailedMessageQueue and 2 in v3::FailedMessageQueue!
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::v2_update_message_queue] InitMessageQueue: Migration done with 0 in total!
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::v2_update_message_queue] InitMessageQueue: Post checks done!
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig PRE Migration counts 4 updated domains and 0 removed domains
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig PRE: Checks done with 4 domains!
[2024-09-03T06:51:11Z INFO  frame_support::migrations] 🚚 Pallet "LiquidityPoolsGateway" VersionedMigration migrating storage version from 2 to 3.
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig ON_RUNTIME_UPGRADE: Migration done with 4 updated domains!
[2024-09-03T06:51:11Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig POST: Checks done!
```

### Centrifuge Chain
```
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Pre checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::nuke] ClearRelayerList: Cleared all RelayerList storage entries
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::nuke] ClearRelayerList Clear result of RelayerList: backend: 0 unique: 0 loops: 0
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Migration done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Post checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::v0_init_message_queue] InitMessageQueue: Pre checks done with 0 items in OutboundMessageQueue and 0 items in FailedOutboundMessages!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::v0_init_message_queue] InitMessageQueue: Initiating migration of 0 outbound messages
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::v0_init_message_queue] InitMessageQueue: Initiating migration of 0 failed outbound messages
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::v0_init_message_queue] InitMessageQueue: Migration done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::v0_init_message_queue] InitMessageQueue: Post checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig PRE Migration counts 4 updated domains and 0 removed domains
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig PRE: Checks done with 4 domains!
[2024-09-03T06:52:12Z INFO  frame_support::migrations] 🚚 Pallet "LiquidityPoolsGateway" VersionedMigration migrating storage version from 0 to 3.
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig ON_RUNTIME_UPGRADE: Migration done with 4 updated domains!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::init_axelar_router] DomainRoutersToAxelarConfig POST: Checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Pre checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::nuke] ClearRelayerList: Cleared all RelayerList storage entries
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::nuke] ClearRelayerList Clear result of RelayerList: backend: 0 unique: 0 loops: 0
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Migration done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::liquidity_pools_v2::kill_relayer_list] ClearRelayerList: Post checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::foreign_investments_v2] ForeignInvestmentsV2: Pre checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::foreign_investments_v2] ForeignInvestmentsV2: Removed 3 ForeignInvestmentInfo V1 keys
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::foreign_investments_v2] ForeignInvestmentsV2: Migration done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::foreign_investments_v2] ForeignInvestmentsV2: Post checks done!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "ForeignInvestments" from StorageVersion(1) to StorageVersion(2)
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::permissions_v1] PermssionsV1: Pre checks done!
[2024-09-03T06:52:12Z INFO  frame_support::migrations] 🚚 Pallet "Permissions" VersionedMigration migrating storage version from 0 to 1.
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::permissions_v1] PermssionsV1: Migrated 34 permissions!
[2024-09-03T06:52:12Z INFO  runtime_common::migrations::permissions_v1] PermssionsV1: Post checks done!
[2024-09-03T06:52:12Z INFO  runtime::frame-support] ⚠️ Ethereum declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `NoStorageVersionSet`
[2024-09-03T06:52:12Z INFO  runtime::executive] ✅ Entire runtime state decodes without error. 6103042 bytes total.
 ```

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: runtime/common/src/migrations/liquidity_pools_gateway.rs
Comment (lemunozm) [inline runtime/common/src/migrations/liquidity_pools_gateway.rs:161]: Could be this: `Option<v0::DomainRouter<T>>`? cc @wischli

Comment (wischli) [reply runtime/common/src/migrations/liquidity_pools_gateway.rs:161]: Unfortunately I had tried that already. Still unable to decode.

File: runtime/centrifuge/src/migrations.rs
Comment (wischli) [inline runtime/centrifuge/src/migrations.rs:35]: FYI: The issue with `try-runtime` decoding errors were the idempocy checks which re-executed the pre- and post-checks. It appears to me that during re-execution, the values were already migrated such that they failed to decode for `v0`. Therefore, I bumped the storage version of the gateway once again and wrapped the migration into a `VersionedMigration`.

**Review by wischli — COMMENTED** [review]
No need to merge this PR as `DomainRouters` has been completely deprecated. Going to use this PR as base for migrating the previous `DomainRouters` to the new settings storage(s).

File: runtime/centrifuge/src/migrations.rs
Comment (lemunozm) [reply runtime/centrifuge/src/migrations.rs:35]: > I bumped the storage version of the gateway once again

That's weird. From 0.14.0, the version was already increment, isn't?

File: runtime/common/src/migrations/liquidity_pools_v2.rs
Comment (wischli) [inline runtime/common/src/migrations/liquidity_pools_v2.rs:100]: Note: Dev and Demo were already upgraded to the intermediary storage version 2 which still supported `gateway::DomainRouters` storage but removed all routers but the `AxelarEvm` ones. The corresponding commit hash is https://github.com/centrifuge/centrifuge-chain/pull/1966/commits/8ec41e92a8b0bd31fef1c90e6a47cbcd48e771e6

Therefore, we need to provide migrations from v0 and v2.

File: runtime/common/src/migrations/liquidity_pools_v2.rs
Comment (wischli) [inline runtime/common/src/migrations/liquidity_pools_v2.rs:424]: Note: There are two `pre_upgrade` hooks in this PR which utilize a transactional layer to ensure the exact migration call which is applied in the corresponding `on_runtime_upgrade` does not fail. However, since those migration calls mutate storage, I am reverting the storage changes by failing the transactional layer on purpose.

File: pallets/liquidity-pools-gateway/src/lib.rs
Comment (cdamian) [inline pallets/liquidity-pools-gateway/src/lib.rs:69]: Can you please tell me why do we jump one version here?

File: runtime/common/src/migrations/liquidity_pools_v2.rs
Comment (cdamian) [inline runtime/common/src/migrations/liquidity_pools_v2.rs:313]: Shouldn't these reads/writes args be references?

File: pallets/liquidity-pools-gateway/src/lib.rs
Comment (wischli) [reply pallets/liquidity-pools-gateway/src/lib.rs:69]: Dev and Demo are using storage version 2 which represents the `DomainRouters` migration in which we removed all key types except for `AxelarEvm`. So to distinguish between the Dev/Demo storage version and the current state of the repo, we need to bump another time.

File: runtime/common/src/migrations/liquidity_pools_v2.rs
Comment (wischli) [reply runtime/common/src/migrations/liquidity_pools_v2.rs:313]: They could be yes. But is there any advantage to gain from that? Where do you see an issue with the current style?

Comment (cdamian) [reply runtime/common/src/migrations/liquidity_pools_v2.rs:313]: In the caller of this fn - https://github.com/centrifuge/centrifuge-chain/pull/1966/files#diff-701a3e1aba84712266444a4d0cdee431c7a71b7cd635255592de9e26f6d77714R252, if you don't provide mutable references they will remain 0?

**Review by mustermeiszer — COMMENTED** [review]
Looks good!

File: runtime/common/src/migrations/liquidity_pools_v2.rs
Comment (wischli) [reply runtime/common/src/migrations/liquidity_pools_v2.rs]: Fixed by [bf97ee8](https://github.com/centrifuge/centrifuge-chain/pull/1966/commits/bf97ee805ed145ed527fe03dd076b427e4e05234)

**Review by cdamian — APPROVED** [review]
Thank you very much for tackling these @wischli!

**Comment by wischli** [comment]
> Looks good!

Would you mind approving? Requiring your approval as code owner.

**Review by mustermeiszer — APPROVED** [review]

**Pull request merged**
